lib/Sema/TypeCheckStmt.cpp

```diff
@@ -1010,6 +1010,35 @@ static bool isDiscardableType(Type type) {
           type->lookThroughAllAnyOptionalTypes()->isVoid());
 }
 
+static void diagnoseIgnoredLiteral(TypeChecker &TC, LiteralExpr *LE) {
+  const auto getLiteralDescription = [](LiteralExpr *LE) -> StringRef {
+    switch (LE->getKind()) {
+    case ExprKind::IntegerLiteral: return "integer";
+    case ExprKind::FloatLiteral: return "floating-point";
+    case ExprKind::BooleanLiteral: return "boolean";
+    case ExprKind::StringLiteral: return "string";
+    case ExprKind::InterpolatedStringLiteral: return "string";
+    case ExprKind::MagicIdentifierLiteral:
+      switch (cast<MagicIdentifierLiteralExpr>(LE)->getKind()) {
+      case MagicIdentifierLiteralExpr::Kind::File: return "#file";
+      case MagicIdentifierLiteralExpr::Kind::Line: return "#line";
+      case MagicIdentifierLiteralExpr::Kind::Column: return "#column";
+      case MagicIdentifierLiteralExpr::Kind::Function: return "#function";
+      case MagicIdentifierLiteralExpr::Kind::DSOHandle: return "#dsohandle";
+      }
+    case ExprKind::NilLiteral:
+    case ExprKind::ObjectLiteral:
+      llvm_unreachable("ignored nil/object literals should not typecheck");
+    default:
```

Comment:
If ever we add a new literal kind, we're going to forget to update this code and crash here. Idea: add a `LITERAL_EXPR` X-macro to `ExprNodes.def` and use it to mark the (non-abstract) literal expressions. Then you can

  #define LITERAL_EXPR(Id, Parent)
  #define ABSTRACT_EXPR(Id, Parent)
  #define EXPR(Id, Parent) case ExprKind::Id:
  #include "swift/AST/ExprNodes.def"
    llvm_unreachable("Not a literal expression");

in your switch statement to get all of the non-literal expression kinds listed, and drop the `default` clause. When someone adds a new literal kind, they'll get a warning on this incomplete switch.